Fix update memory without memoryswap

coolljt0725 opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**\- What I did**

fixes #25460 

**\- How I did it**

**\- How to verify it**

**\- Description for the changelog**

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

**\- A picture of a cute animal (not mandatory but encouraged)**

The memory should always be smaller than memoryswap,
we should error out with message that user know how
to do rather than just an invalid argument error if
user update the memory limit bigger than already set
memory swap.

Signed-off-by: Lei Jitang leijitang@huawei.com

justincormack commented on man/docker-update.1.md:
```diff
@@ -67,6 +67,12 @@ a running container with kernel memory initialized.
+   If you want update a memory limit bigger than the already set swap memory limit,
+   you should update swap memroy limit at the same time. If you don't set swap memory 
+   limit on docker create/run but only memory limit, the swap memory is double of
+   the memory limit.
```
remove the "of" here, or use "double that of", it reads a bit odd.

coolljt0725 replied:
fixed, thanks :)

justincormack commented on man/docker-update.1.md:
```diff
@@ -67,6 +67,12 @@ a running container with kernel memory initialized.
 
+   Note that the memory should be smaller than the already set swap memory limit.
+   If you want update a memory limit bigger than the already set swap memory limit,
+   you should update swap memroy limit at the same time. If you don't set swap memory 
```
typo "memroy"

coolljt0725 replied:
fixed

coolljt0725 committed:
Fix update memory without memoryswap

The memory should always be smaller than memoryswap,
we should error out with message that user know how
to do rather than just an invalid argument error if
user update the memory limit bigger than already set
memory swap.

Signed-off-by: Lei Jitang <leijitang@huawei.com>

justincormack commented:
LGTM

thaJeztah commented on container/container_unix.go:
```diff
@@ -283,6 +283,11 @@ func (container *Container) UpdateContainer(hostConfig *containertypes.HostConfi
 	if resources.Memory != 0 {
+		// if memory limit smaller than already set memoryswap limit and doesn't
+		// update the memoryswap limit, then error out.
+		if resources.Memory > cResources.MemorySwap && resources.MemorySwap == 0 {
```
Should this be 

```
if resources.Memory > cResources.MemorySwap &&  resources.Memory > resources.MemorySwap {
```

To check if a *new* MemorySwap is set, it's a valid new combination as well?

thaJeztah replied:
er, sorry, my example wasn't right, hold on

thaJeztah replied:
think this;

```golang
if resources.Memory > cResources.MemorySwap &&  (resources.MemorySwap == 0 || resources.Memory > resources.MemorySwap) {
```

thaJeztah replied:
LOL, actually, my first example may be correct

coolljt0725 replied:
@thaJeztah  We already has `resources.Memory > resources.MemorySwap` validation before we get here in https://github.com/docker/docker/blob/master/daemon/daemon_unix.go#L273. So I think it's ok just check if the memoryswap is set here.

thaJeztah replied:
@coolljt0725 ah, makes sense, thanks!

thaJeztah commented:
LGTM
